editor.planx.uk/src/hooks/useSearch.ts

```diff
@@ -0,0 +1,53 @@
+import Fuse, { IFuseOptions } from "fuse.js";
+import { useEffect, useMemo, useState } from "react";
+
+interface UseSearchProps<T extends object> {
+  list: T[];
+  keys: string[];
+}
+
+export interface SearchResult<T extends object> {
+  item: T;
+  key: string;
+  matchIndices?: [number, number][];
+}
+
+export type SearchResults<T extends object> = SearchResult<T>[];
+
+export const useSearch = <T extends object>({
+  list,
+  keys,
+}: UseSearchProps<T>) => {
+  const [pattern, setPattern] = useState("");
+  const [results, setResults] = useState<SearchResults<T>>([]);
+
+  const fuseOptions: IFuseOptions<T> = useMemo(
+    () => ({
+      useExtendedSearch: true,
```

Comment:
To clarify these Fuse.js configs, I've read the docs so just looking to confirm...

- useExtendedSearch
- - _This is used to allow tooling for exact searching, and statements, and other operators?_
- includeMatches
- - _Gives you the location of the matching text?_

Would we want to sort the results by the Fuse score? Don't know if we've discussed this already... looks like Fuse has that scoring built in if we did want to.

Reply:
That's correct 👍 

- `useExtendedSearch` gives us a few additional options for searching, e.g. starts with, ends with, exact etc
- `includeMatches` is required to give us indices to use for highlighting the results in bold

Fuse.js is sorting the results by score already -`shouldSort` is `true by default - ([docs](https://www.fusejs.io/api/options.html#shouldsort))).